athena/models/fastspeech.py

```diff
@@ -0,0 +1,404 @@
+                padding="same",
+                use_bias=False,
+                data_format="channels_last",
+            )(inner)
+            if self.hparams.batch_norm_position == 'after':
+                inner = tf.nn.tanh(inner)
+                inner = layers.BatchNormalization()(inner)
+            else:
+                inner = layers.BatchNormalization()(inner)
+                inner = tf.nn.tanh(inner)
+            inner = layers.Dropout(self.hparams.dropout_rate)(inner)
+        inner = layers.Dense(self.feat_dim, name='projection')(inner)
+        self.postnet = tf.keras.Model(inputs=input_features_postnet, outputs=inner, name="postnet")
+        print(self.postnet.summary())
+
+        self.teacher_model = None
+        self.teacher_type = None
+        self.duration_calculator = DurationCalculator()
+
+        # define loss function and metrics
+        self.loss_function = FastSpeechLoss(self.hparams.duration_predictor_loss_weight,
+                                            teacher_guide=self.hparams.teacher_guide_loss)
+        self.metric = tf.keras.metrics.Mean(name="AverageLoss")
+
+    def set_teacher_model(self, teacher_model, teacher_type):
+        self.teacher_model = teacher_model
+        self.teacher_type = teacher_type
+        self.duration_calculator = DurationCalculator(self.teacher_model, self.teacher_type)
+
+    def restore_from_pretrained_model(self, pretrained_model, model_type=""):
+        if model_type == "":
+            return
+        if model_type == "tts_transformer":
+            logging.info("loading from pretrained tts transformer model")
+            self.x_net = pretrained_model.x_net
+            self.encoder = pretrained_model.encoder
+        else:
+            raise ValueError("NOT SUPPORTED")
+
+    def get_loss(self, outputs, samples, training=None):
+        loss = self.loss_function(outputs, samples)
+        total_loss = sum(list(loss.values())) if isinstance(loss, dict) else loss
+        self.metric.update_state(total_loss)
+        metrics = {self.metric.name: self.metric.result()}
+        return loss, metrics
+
+    def _feedforward_decoder(self, encoder_output, duration_indexes, duration_sequences,
+                             output_length, training):
+        if training is True:
+            expanded_array = self.length_regulator(encoder_output, duration_indexes,
+                                                   output_length=output_length)
+            expanded_length = output_length
+        else:
+            expanded_array = self.length_regulator.inference(encoder_output, duration_sequences,
+                                                             alpha=self.hparams.alpha)
+            expanded_length = tf.reduce_sum(duration_sequences, axis=1) # [batch]
+        expanded_array = tf.ensure_shape(expanded_array, [None, None, self.hparams.d_model])
+        expanded_mask, _ = SpeechTransformer._create_masks(expanded_array, expanded_length, None)
```

Comment:
We should find a way to combine SpeechTransformer._create_masks() and TTSTransformer._create_masks() together and put the _create_masks method into utils

Reply:
I will push another request for this update.